### ISSUE TITLE
Cannot use a function that returns a type in a type annotation

### ISSUE THREAD

**Issue by rdong8** [microsoft/pyright#9501]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

If you have a question about typing or a behavior that you’re seeing in Pyright (as opposed to a bug report or enhancement request), consider posting to the [Pyright discussion forum](https://github.com/microsoft/pyright/discussions).

**Describe the bug**
A clear and concise description of the behavior you are seeing and the expected behavior along with steps to reproduce it.

The result of a function call that returns a type is not treated as a type, and thus cannot be used as a type annotation, even when pyright knows that the value is a type. 

**Code or Screenshots**
If possible, provide a minimal, self-contained code sample (surrounded by triple back ticks) to demonstrate the issue. The code should define or import all referenced symbols.

```python
def give_type() -> type:
    return int


def f(x: give_type()):  # Error
    pass


T = give_type()


def g(x: T):  # Error
    pass
```

If your code relies on symbols that are imported from a third-party library, include the associated import statements and specify which versions of those libraries you have installed.

**VS Code extension or command-line**
Are you running pyright as a VS Code extension, a language server in another editor, integrated into Pylance, or the command-line tool? Which version?

Extension.

For more context, I'm trying to make dependency injection in FastAPI less cumbersome.

```python
from __future__ import annotations
import contextlib
from typing import Any, AsyncGenerator

from fastapi import FastAPI
from fastapi.params import Depends
from typing_extensions import Annotated


class DB:
    @classmethod
    async def create(cls) -> DB:
        return cls()

    async def close(self) -> None:
        pass


dependencies: dict[type, Any] = {}


def depends[Dependency](dependency: type[Dependency]) -> Annotated:
    return Annotated[Dependency, Depends(lambda: dependencies[dependency])]


def set_dependency(dependency: Any) -> None:
    dependencies[type(dependency)] = dependency


DBDependency = depends(DB)


@contextlib.asynccontextmanager
async def lifespan(app: FastAPI) -> AsyncGenerator[None, None]:
    # Startup tasks
    db = await DB.create()
    set_dependency(db)

    yield

    # Shutdown tasks
    await db.close()
    print("SHUTDOWN")


app = FastAPI(lifespan=lifespan)


@app.get("/foo")
async def foo(
    db: DBDependency,  # Allows you to write this instead of `db: Annotated[DB, Depends(lambda: dependencies[DB])]`
) -> dict[str, str]:
    # Can use db here
    return {"message": "Hello World"}
```

This runs but doesn't pass type checking.

**Comment by erictraut** [comment]
What you're doing here violates the Python typing spec, so pyright is correct in generating an error here. While your code will run, it will not pass static type checking. If you want your code to work with static type checkers, you'll need to follow the rules of the type system. In particular, you cannot use variables in type expressions. See [this section](https://typing.readthedocs.io/en/latest/spec/annotations.html) of the Python typing spec for details.
